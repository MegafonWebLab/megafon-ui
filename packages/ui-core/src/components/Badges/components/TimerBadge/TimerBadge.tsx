import React, { useEffect, useState } from 'react';
import { cnCreate, filterDataAttrs } from '@megafon/ui-helpers';
import TimerIcon from '@megafon/ui-icons/system-16-timer_16.svg';
import * as PropTypes from 'prop-types';
import Timer from './Timer/Timer';
import './TimerBadge.less';

const SECONDS_IN_HOUR = 3600;
const MS_IN_SECOND = 1000;
export const TIMEOUT = 1000;

const TimerBadgeTheme = {
    RED: 'red',
    GREY: 'grey',
} as const;

export interface ITimerBadgeProps {
    /** Дата окончания таймера */
    expirationDate: Date;
    /** Граница переключения в режим обратного отсчёта, в секундах */
    countdownStart?: number;
    /** Отображение дополнительного текста */
    hasPrefix?: boolean;
    /** Текст перед таймером в режиме обратного отсчета. Появляется при hasPrefix=true. */
    countdownText?: string;
    /** Текст перед таймером в режиме даты. Появляется при hasPrefix=true. */
    expirationDateText?: string;
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Дополнительные data-атрибуты к внутренним элементам */
    dataAttrs?: {
        root?: Record<string, string>;
        text?: Record<string, string>;
        timer?: Record<string, string>;
        iconContainer?: Record<string, string>;
    };
}

const cn = cnCreate('mfui-timer-badge');
const TimerBadge: React.FC<ITimerBadgeProps> = ({
    expirationDate,
    countdownStart = 86400,
    hasPrefix = false,
    countdownText = 'Закончится через',
    expirationDateText = 'Воспользуйтесь до',
    className,
    dataAttrs,
}) => {
    const remainingTime = (expirationDate.getTime() - Date.now()) / MS_IN_SECOND;

    const [actualRemainingTime, setActualRemainingTime] = useState(remainingTime);
    const [showCountdown, setShowCountdown] = useState(false);

    const isLastHour = actualRemainingTime < SECONDS_IN_HOUR;
    const isTimeExpired = actualRemainingTime < 1;

    const theme = showCountdown ? TimerBadgeTheme.RED : TimerBadgeTheme.GREY;
    const additionalText = showCountdown ? countdownText : expirationDateText;

    useEffect(() => {
        setShowCountdown(actualRemainingTime < countdownStart);
    }, [actualRemainingTime, countdownStart]);

    useEffect(() => {
        if (isTimeExpired) {
            return undefined;
        }

        const timeoutId = window.setTimeout(() => setActualRemainingTime(actualRemainingTime - 1), TIMEOUT);

        return (): void => clearTimeout(timeoutId);
    }, [isTimeExpired, showCountdown, actualRemainingTime]);

    return (
        <div {...filterDataAttrs(dataAttrs?.root)} className={cn({ theme }, className)}>
            <div
                {...filterDataAttrs(dataAttrs?.iconContainer)}
                className={cn('icon-container', { shadow: isLastHour && showCountdown })}
            >
                <TimerIcon className={cn('icon')} />
            </div>
            <div {...filterDataAttrs(dataAttrs?.text)} className={cn('text')}>
                {isTimeExpired ? (
                    'Время действия истекло'
                ) : (
                    <Timer
                        additionalText={hasPrefix ? additionalText : undefined}
                        showCountdown={showCountdown}
                        expirationDate={expirationDate}
                        actualRemainingTime={actualRemainingTime}
                        dataAttrs={{ root: dataAttrs?.timer }}
                    />
                )}
            </div>
        </div>
    );
};

TimerBadge.propTypes = {
    expirationDate: PropTypes.instanceOf(Date).isRequired,
    countdownStart: PropTypes.number,
    hasPrefix: PropTypes.bool,
    countdownText: PropTypes.string,
    expirationDateText: PropTypes.string,
    className: PropTypes.string,
    dataAttrs: PropTypes.shape({
        root: PropTypes.objectOf(PropTypes.string.isRequired),
        text: PropTypes.objectOf(PropTypes.string.isRequired),
        timer: PropTypes.objectOf(PropTypes.string.isRequired),
        iconContainer: PropTypes.objectOf(PropTypes.string.isRequired),
    }),
};

export default TimerBadge;
