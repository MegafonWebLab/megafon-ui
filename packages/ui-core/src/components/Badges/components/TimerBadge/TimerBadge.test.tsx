import * as React from 'react';
import { render, act } from '@testing-library/react';
import TimerBadge, { TIMEOUT } from './TimerBadge';

const currentDate = new Date('2022-03-15T10:50:15z');
const lessThanHourBeforeEndDate = new Date('2022-03-15T11:40:15z');
const fiveSecondsBeforeEndDate = new Date('2022-03-15T10:50:20z');
const fiveDaysBeforeEndDate = new Date('2022-03-20T14:50:15z');
const expiredDate = new Date('2022-03-15T09:50:15z');
const sixDays = 600000;
const fourDays = 400000;

const dataAttrs = {
    root: { 'data-testid': 'root' },
    text: { 'data-testid': 'text' },
    timer: { 'data-testid': 'timer' },
    iconContainer: { 'data-testid': 'iconContainer' },
};

describe('<TimerBadge />', () => {
    const defaultDateNow = Date.now;

    beforeAll(() => {
        Date.now = jest.fn().mockReturnValue(currentDate.getTime());
    });

    afterAll(() => {
        Date.now = defaultDateNow;
    });

    it('should render with require props', () => {
        const { container } = render(<TimerBadge expirationDate={fiveDaysBeforeEndDate} />);

        expect(container).toMatchSnapshot();
    });

    it('should render with dataAttrs', () => {
        const { getByTestId } = render(<TimerBadge expirationDate={fiveDaysBeforeEndDate} dataAttrs={dataAttrs} />);

        expect(getByTestId('root')).toBeTruthy();
        expect(getByTestId('text')).toBeTruthy();
        expect(getByTestId('timer')).toBeTruthy();
        expect(getByTestId('iconContainer')).toBeTruthy();
    });

    it('should render with className', () => {
        const { getByTestId } = render(
            <TimerBadge expirationDate={fiveDaysBeforeEndDate} className="custom-class" dataAttrs={dataAttrs} />,
        );

        expect(getByTestId('root')).toHaveClass('custom-class');
    });

    it('should render when expirationDate expired', () => {
        const { queryByText } = render(<TimerBadge expirationDate={expiredDate} dataAttrs={dataAttrs} />);

        expect(queryByText('Время действия истекло')).toBeInTheDocument();
    });

    it('should render when countdownStart is greater than expirationDate and hasPrefix is false', () => {
        const { getByTestId, queryByText } = render(
            <TimerBadge
                countdownStart={sixDays}
                countdownText="Срок действия истечёт через"
                expirationDate={fiveDaysBeforeEndDate}
                dataAttrs={dataAttrs}
                hasPrefix={false}
            />,
        );

        expect(getByTestId('root')).toHaveClass('mfui-timer-badge_theme_red');
        expect(queryByText('Срок действия истечёт через')).not.toBeInTheDocument();
        expect(queryByText('5 дн 4 ч')).toBeTruthy();
    });

    it('should render when countdownStart is greater than expirationDate and hasPrefix is true', () => {
        const { getByTestId, queryByText } = render(
            <TimerBadge
                countdownStart={sixDays}
                countdownText="Срок действия истечёт через"
                expirationDate={fiveDaysBeforeEndDate}
                dataAttrs={dataAttrs}
                hasPrefix
            />,
        );

        expect(getByTestId('root')).toHaveClass('mfui-timer-badge_theme_red');
        expect(queryByText('Срок действия истечёт через')).toBeInTheDocument();
        expect(queryByText('5 дн 4 ч')).toBeTruthy();
    });

    it('should render when countdownStart is less than expirationDate and hasPrefix is false', () => {
        const { getByTestId, queryByText } = render(
            <TimerBadge
                countdownStart={fourDays}
                expirationDate={fiveDaysBeforeEndDate}
                expirationDateText="Действительно до"
                dataAttrs={dataAttrs}
                hasPrefix={false}
            />,
        );

        expect(getByTestId('root')).toHaveClass('mfui-timer-badge_theme_grey');
        expect(queryByText('Действительно до')).not.toBeInTheDocument();
        expect(queryByText('20 марта 2022')).toBeTruthy();
    });

    it('should render when countdownStart is less than expirationDate and hasPrefix is true', () => {
        const { getByTestId, queryByText } = render(
            <TimerBadge
                countdownStart={fourDays}
                expirationDate={fiveDaysBeforeEndDate}
                expirationDateText="Действительно до"
                dataAttrs={dataAttrs}
                hasPrefix
            />,
        );

        expect(getByTestId('root')).toHaveClass('mfui-timer-badge_theme_grey');
        expect(queryByText('Действительно до')).toBeInTheDocument();
        expect(queryByText('20 марта 2022')).toBeTruthy();
    });

    it('should render when expirationDate is less than hour and countdownStart is greater than expirationDate', () => {
        const { getByTestId, queryByText } = render(
            <TimerBadge countdownStart={sixDays} expirationDate={lessThanHourBeforeEndDate} dataAttrs={dataAttrs} />,
        );

        expect(getByTestId('iconContainer')).toHaveClass('mfui-timer-badge__icon-container_shadow');
        expect(queryByText('50 мин')).toBeTruthy();
    });

    it('should render when expirationDate decremented by 1 after rerender', () => {
        jest.useFakeTimers();

        const { queryByText, rerender } = render(
            <TimerBadge expirationDate={fiveSecondsBeforeEndDate} dataAttrs={dataAttrs} />,
        );

        expect(queryByText('5 сек')).toBeTruthy();

        act(() => {
            jest.runTimersToTime(TIMEOUT);
        });

        rerender(
            <TimerBadge countdownStart={sixDays} expirationDate={fiveSecondsBeforeEndDate} dataAttrs={dataAttrs} />,
        );

        expect(queryByText('4 сек')).toBeTruthy();

        jest.useRealTimers();
    });
});
