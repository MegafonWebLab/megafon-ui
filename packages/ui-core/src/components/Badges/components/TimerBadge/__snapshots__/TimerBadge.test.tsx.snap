// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimerBadge /> should render with require props 1`] = `
<div>
  <div
    class="mfui-timer-badge mfui-timer-badge_theme_grey"
  >
    <div
      class="mfui-timer-badge__icon-container"
    >
      <test-file-stub
        classname="mfui-timer-badge__icon"
      />
    </div>
    <div
      class="mfui-timer-badge__text"
    >
      <div
        class="mfui-timer"
      >
        <span>
          20 марта 2022
        </span>
      </div>
    </div>
  </div>
</div>
`;
